File: lib/php/examples/01_provision_peer.php
#!/usr/bin/env php
<?php
/**
 * 01_provision_peer — Create a peer for a user, write the .conf and QR code to disk.
 *
 * Usage:
 *   WG_BASE_URL=http://... WG_ADMIN_USER=... WG_API_TOKEN=... php examples/01_provision_peer.php
 */

declare(strict_types=1);

require_once __DIR__ . '/../vendor/autoload.php';

use WgPortal\Client\{WgPortalClient, WgPortalException};

function requireEnv(string $key): string
{
    $value = getenv($key);
    if ($value === false || $value === '') {
        fwrite(STDERR, "Missing required env var: {$key}\n");
        exit(1);
    }
    return $value;
}

$client = new WgPortalClient(
    requireEnv('WG_BASE_URL'),
    requireEnv('WG_ADMIN_USER'),
    requireEnv('WG_API_TOKEN')
);

$email = 'alice@example.com';

// Resolve or create user
try {
    $info = $client->provisioning()->getUserInfo(null, $email);
    $userId = $info['UserIdentifier'];
    echo "User exists: {$userId}\n";
} catch (WgPortalException $e) {
    if ($e->getHttpCode() === 404) {
        echo "Creating user {$email} ...\n";
        $user = $client->users()->create([
            'Identifier' => 'alice',
            'Email' => $email,
            'Firstname' => 'Alice',
            'Password' => 'changeme123',
        ]);
        $userId = $user['Identifier'];
        echo "Created user: {$userId}\n";
    } else {
        fwrite(STDERR, "Error: {$e->getMessage()}\n");
        exit(1);
    }
}

// Get active interface
$interfaces = $client->interfaces()->getAll();
$active = array_values(array_filter($interfaces, fn(array $i) => !($i['Disabled'] ?? false)));

if (empty($active)) {
    fwrite(STDERR, "No active interfaces\n");
    exit(1);
}
$ifaceId = $active[0]['Identifier'];

// Provision peer
$peer = $client->provisioning()->newPeer([
    'InterfaceIdentifier' => $ifaceId,
    'UserIdentifier' => $userId,
    'DisplayName' => "{$email} - laptop",
]);
echo "Peer: {$peer['Identifier']}\n";

// Save artifacts
$config = $client->provisioning()->getPeerConfig($peer['Identifier']);
$qrPng = $client->provisioning()->getPeerQrCode($peer['Identifier']);

$outDir = __DIR__ . "/../output/{$peer['Identifier']}";
mkdir($outDir, 0755, true);
file_put_contents("{$outDir}/peer.conf", $config);
file_put_contents("{$outDir}/qr.png", $qrPng);
echo "Saved to {$outDir}\n";

File: lib/python/examples/01_provision_peer.py
#!/usr/bin/env python3
"""
01_provision_peer.py — Create a peer for a user, write the .conf file and QR code to disk.

Usage:
  WG_BASE_URL=http://... WG_ADMIN_USER=... WG_API_TOKEN=... python 01_provision_peer.py
"""
import os
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))
from wg_portal_client import WgPortalClient, WgPortalError

CLIENT = WgPortalClient(
    base_url=os.getenv("WG_BASE_URL", "http://192.168.1.100:8888"),
    username=os.getenv("WG_ADMIN_USER", "admin@wgportal.local"),
    api_token=os.getenv("WG_API_TOKEN", ""),
)


def main():
    # Resolve or create user
    email = "alice@example.com"
    try:
        info = CLIENT.provisioning.get_user_info(email=email)
        user_id = info["UserIdentifier"]
        print(f"User exists: {user_id}")
    except WgPortalError:
        user = CLIENT.users.create({
            "Identifier": "alice",
            "Email": email,
            "Firstname": "Alice",
            "Password": "changeme123",
        })
        user_id = user["Identifier"]
        print(f"Created user: {user_id}")

    # Get active interface
    interfaces = CLIENT.interfaces.get_all()
    active = [i for i in interfaces if not i.get("Disabled")]
    if not active:
        print("No active interfaces")
        sys.exit(1)
    iface_id = active[0]["Identifier"]

    # Provision peer
    peer = CLIENT.provisioning.new_peer({
        "InterfaceIdentifier": iface_id,
        "UserIdentifier": user_id,
        "DisplayName": f"{email} - laptop",
    })
    print(f"Peer: {peer['Identifier']}")

    # Save artifacts
    config = CLIENT.provisioning.get_peer_config(peer["Identifier"])
    qr_png = CLIENT.provisioning.get_peer_qr_code(peer["Identifier"])

    out_dir = Path.cwd() / "output" / peer["Identifier"]
    out_dir.mkdir(parents=True, exist_ok=True)
    (out_dir / "peer.conf").write_text(config)
    (out_dir / "qr.png").write_bytes(qr_png)
    print(f"Saved to {out_dir}")


if __name__ == "__main__":
    main()
